cdap-messaging-spi/src/main/java/io/cdap/cdap/messaging/spi/MessagingService.java

```diff
@@ -142,10 +151,18 @@ void rollback(TopicId topicId, RollbackDetail rollbackDetail)
    * system.
    *
    * @param messageFetchRequest the request for fetching messages
-   * @throws TopicNotFoundException if the topic does not exist
-   * @throws IOException if it fails to create the iterator
+   * @throws TopicNotFoundException      if the topic does not exist
+   * @throws IOException                 if it fails to create the iterator
    * @throws ServiceUnavailableException if the messaging service is not available
    */
   CloseableIterator<RawMessage> fetch(MessageFetchRequest messageFetchRequest)
       throws TopicNotFoundException, IOException;
+
+  /**
+   * Cleans up initialized resources. It will only be called once for the lifetime of the messaging
+   * service.
+   *
+   * @param messagingContext messaging service context
+   */
+  void destroy(MessagingContext messagingContext);
```

Comment:
Why this? When do you want to close spanner? We probably should keep the client open in spanner messaging service in appfabric, otherwise it'll hurt the performance substantially.